src/sage/rings/polynomial/integer_valued_polynomials.py

```diff
@@ -0,0 +1,1144 @@
+            except TypeError as exc:
+                msg = 'not a polynomial with integer'
+                msg += f' values: {top_coeff}'
+                raise ValueError(msg) from exc
+            remain += -top_coeff * binomial(N + x, N)
+            result += top_coeff * B[N]
+        return result
+
+    def from_h_vector(self, h):
+        """
+        Convert from some `h`-vector.
+
+        INPUT:
+
+        - ``h`` -- a tuple or vector
+
+        EXAMPLES::
+
+            sage: A = IntegerValuedPolynomialRing(ZZ)
+            sage: B = A.basis()
+            sage: ex = B[2]+B[4]
+            sage: A.from_h_vector(ex.h_vector())
+            B[2] + B[4]
+        """
+        B = self.basis()
+        d = len(h) - 1
+        m = matrix(QQ, d + 1, d + 1,
+                   lambda j, i: (-1)**(d - j) * binomial(d - i, d - j))
+        v = vector(QQ, [h[i] for i in range(d + 1)])
+        return self.sum(Integer(c) * B[i] for i, c in enumerate(m * v))
+
+    def gen(self):
+        r"""
+        Return the generator of the algebra.
+
+        EXAMPLES::
+
+            sage: F = IntegerValuedPolynomialRing(ZZ)
+            sage: F.gen()
+            B[1]
+        """
+        return self.algebra_generators()[0]
+
+    @cached_method
+    def algebra_generators(self):
+        r"""
+        Return the generators of this algebra.
+
+        EXAMPLES::
+
+            sage: A = IntegerValuedPolynomialRing(ZZ); A
+            Integer-Valued Polynomial Ring over Integer Ring
+            sage: A.algebra_generators()
+            Family (B[1],)
+        """
+        NonNeg = self.basis().keys()
+        return Family([self.monomial(NonNeg(1))])
+
+    gens = algebra_generators
+
+    def _element_constructor_(self, x):
+        r"""
+        Convert ``x`` into ``self``.
+
+        INPUT:
+
+        - ``x`` -- an integer or something convertible
+
+        EXAMPLES::
+
+            sage: R = IntegerValuedPolynomialRing(QQ)
+            sage: x = R.gen()
+            sage: R(3)
+            3*B[0]
+            sage: R(x)
+            B[1]
+        """
+        if x in NonNegativeIntegers():
+            W = self.basis().keys()
+            return self.monomial(W(x))
+
+        P = x.parent()
+        if isinstance(P, IntegerValuedPolynomialRing):
+            if P is self:
+                return x
+            if P is not self.base_ring():
+                return self.element_class(self, x.monomial_coefficients())
+
+        # ok, not a integer-valued polynomial ring element (or should not be
+        # viewed as one).
+        if isinstance(x, str):
+            from sage.misc.sage_eval import sage_eval
+            return sage_eval(x, locals=self.gens_dict())
+        R = self.base_ring()
+        # coercion via base ring
+        x = R(x)
+        if x == 0:
+            return self.element_class(self, {})
+        return self.from_base_ring_from_one_basis(x)
+
+    def _coerce_impl(self, x):
```

Comment:
We aren't using the old coercion framework here, so is this actually called? Shouldn't this be put into `_coerce_map_from_` instead?

Reply:
I have removed this and it works. Thanks for the suggestion.